### PR TITLE
State docfix

### LINKED ISSUE #393: Improve State documentation
Hi,

I think the documentation of states and the way they're used by engines is not detailed enough.
Currently the only way to know what attributes are available in the default State object, used by the engines generated by convenience methods such as `create_supervised_trainer`, is to look at the source code.  
Some of them are mentioned in the [Concept](https://pytorch.org/ignite/concepts.html#state) part of the documentation but there is still a lot of room for confusion about the way they work and some are missing.
For example, no information about the `batch` attribute which without documentation some might assume to correspond to the batch index (kinda like `epoch` for the epoch index) which is actually `iteration`. Same thing for `epoch` which is used implicitly in examples but no mention of it in the default `State` documentation, and you might think it's obvious but then someone will try to create a default `State` and assume `epoch` is a built-in attribute set to 0 at instantiation when in fact you'll have to search through the source code to figure out that it's not a default attribute and is only created by the default `Engine` through a kwarg passed to the `State` class.

Additionally this kind of explicit documentation of the state should be done every time a custom `State` is created otherwise the user is left to guess what attributes are available in it, what they mean and what format they're in. For example, is `output` the return from the forward pass of your model? The loss?
You'll have to look at `Engine` then find out it's the output of the process_function argument. Then figure out that in the case of the  `create_supervised_trainer` that function is `_update` which returns `loss.item()` and in the case of `create_supervised_evaluator` returns `(y_pred, y)`.

Thank you, and sorry for the verbosity but without it I was afraid this would be rejected as trivial.

### PR DESCRIPTION
**Pull request by anmolsjoshi**
Fixes #393  

Description:
Updates concepts.rst and descriptions of create_supervised_trainer and create_supervised_evaluator to provide explanations and examples of engine.state.output

Check list:
* [ ] New tests are added (if a new feature is modified)
* [x] New doc strings: text and/or example code are in RST format
* [x] Documentation is updated (if required)

### CODE REVIEW

File: docs/source/concepts.rst [hunk 1]
```diff
@@ -116,9 +130,66 @@ completed iterations and epochs:
 
     trainer.add_event_handler(Events.ITERATION_COMPLETED, on_iteration_completed)
 
+There are many ways of creating a `process_function`, this is why Ignite provides output_transform parameter for its
```
Comment (vfdev-5) [inline docs/source/concepts.rst]: I would rephrase this as 
```suggestion
Since there is no restrictions on `process_function`, Ignite provides `output_transform` argument for its 
```
@anmolsjoshi what do you think ?

File: docs/source/concepts.rst [hunk 2]
```diff
@@ -116,9 +130,66 @@ completed iterations and epochs:
 
     trainer.add_event_handler(Events.ITERATION_COMPLETED, on_iteration_completed)
 
+There are many ways of creating a `process_function`, this is why Ignite provides output_transform parameter for its 
+:class:`~ignite.metrics` and :class:`~ignite.handlers`. `output_transform` is a function used to transform `engine.state.output` for intended use. Below we'll see different types of `engine.state.output` and how to transform them.
```
Comment (vfdev-5) [inline docs/source/concepts.rst]:
```suggestion
:class:`~ignite.metrics` and :class:`~ignite.handlers`. Argument `output_transform` is a function used to transform `engine.state.output` for intended use. Below we'll see different types of `engine.state.output` and how to transform them.
```

**Comment by vfdev-5** [comment]
Visdom is [broken](https://github.com/facebookresearch/visdom/issues/548)

File: docs/source/concepts.rst (hunk 1 above)
Comment (anmolsjoshi) [reply docs/source/concepts.rst]: It reads better!

File: docs/source/concepts.rst (hunk 2 above)
Comment (anmolsjoshi) [reply docs/source/concepts.rst]: Reads better!

**Comment by anmolsjoshi** [comment]
@vfdev-5 thanks! I've incorporated your comments. please review and merge.
